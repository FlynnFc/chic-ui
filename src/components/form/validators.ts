import validator from 'validator';

export const composeValidators =
  (...validators: any[]) =>
  (value: any, allValues: any) =>
    validators.reduce(
      (error, validator) => error || validator(value, allValues),
      undefined
    );

export const isRequired = (value: any) => {
  const trimmedValue = typeof value === 'string' ? value?.trim() : value;

  if (!trimmedValue && trimmedValue !== 0) {
    return 'Please complete this field.';
  }

  return undefined;
};

export const isEmail = (message: string) => (value: string) => {
  if (validator.isEmail(value)) {
    return undefined;
  } else {
    return message;
  }
};

export const isStrongPassword = (message: string) => (value: string) => {
  const options = {
    minLength: 8,
    minLowercase: 1,
    minUppercase: 1,
    minNumbers: 1,
    minSymbols: 1,
    returnScore: false
  };
  // @ts-ignore
  if (validator.isStrongPassword(value, [options])) {
    return undefined;
  } else {
    return message;
  }
};

