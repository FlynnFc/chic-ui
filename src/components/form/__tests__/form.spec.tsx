import Form from '..';
import { render } from '../../../test/utils';
import { isEmail, isStrongPassword } from '../validators';

const onSubmit = (values: any) => {
  console.log('submit form: ', values);
};

describe('Form Component', () => {
  it('Renders Correctly', () => {
    const FormGroup = render(
      <>
        <Form onSubmit={onSubmit}>
          <Form.Input
            placeholder="Enter your full name"
            name="fullname"
            label="Full name"
            required
          />

          <Form.Input
            placeholder="Enter your email"
            name="email"
            label="Email"
            required
            validate={[isEmail('Invalid email.')]}
          />

          <Form.Input
            type="password"
            placeholder="Enter your password"
            name="password"
            label="Password"
            required
            validate={[
              isStrongPassword(
                'Password must be at least 8 characters, 1 lower case characters, 1 upper case characters, 1 numbers and 1 symbols'
              )
            ]}
          />

          <Form.RadioGroup label="Gender" name="gender">
            <Form.Radio value="male" label="Male" />
            <Form.Radio value="female" label="Female" />
          </Form.RadioGroup>

          <Form.Input
            label="Description"
            name="bio"
            placeholder="Enter your bio"
            rows={5}
          />

          <Form.CheckboxGroup label="Internet" required>
            <Form.Checkbox label="3G" value="3g" name="internet" id="3g" />
            <Form.Checkbox label="4G" value="4g" name="internet" id="4g" />
          </Form.CheckboxGroup>

          <Form.SubmitButton>Submit</Form.SubmitButton>
        </Form>
      </>
    );
    expect(FormGroup).toMatchSnapshot();
  });
});
