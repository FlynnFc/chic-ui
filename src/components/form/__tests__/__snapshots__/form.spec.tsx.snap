// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component Renders Correctly 1`] = `
<form
  autoComplete="off"
  onSubmit={[Function]}
>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="fullname"
    >
      Full name
    </label>
    <input
      className="sc-hKgJUU gJnbEf"
      id="fullname"
      name="fullname"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter your full name"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="sc-hKgJUU gJnbEf"
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter your email"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="password"
    >
      Password
    </label>
    <input
      className="sc-hKgJUU gJnbEf"
      id="password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter your password"
      type="password"
      value=""
    />
  </div>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="gender"
    >
      Gender
       (Optional)
    </label>
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="sc-gsTEea ldJggX"
      >
        <label
          className="sc-eCstlR djNJYa"
        >
          <input
            checked={false}
            className="sc-jSgvzq boRgPJ"
            name="gender"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="radio"
            value="male"
          />
          <span />
          Male
        </label>
      </div>
      <div
        className="sc-gsTEea ldJggX"
      >
        <label
          className="sc-eCstlR djNJYa"
        >
          <input
            checked={false}
            className="sc-jSgvzq boRgPJ"
            name="gender"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="radio"
            value="female"
          />
          <span />
          Female
        </label>
      </div>
    </div>
  </div>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="bio"
    >
      Description
       (Optional)
    </label>
    <textarea
      className="sc-hKgJUU fEXmGq"
      id="bio"
      name="bio"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter your bio"
      rows={5}
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-gsTEea ldJggX"
  >
    <label
      className="sc-dlfnuX caUESl"
      htmlFor="chk_group"
    >
      Internet
    </label>
    <div
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div
        className="sc-gsTEea ldJggX"
      >
        <div
          style={
            Object {
              "marginBottom": "15px",
            }
          }
        >
          <input
            checked={false}
            className="sc-iBPTik gPelDo"
            id="3g"
            name="internet"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
            value="3g"
          />
          <label
            className="sc-gKseQn gHGxWx"
            htmlFor="3g"
          >
            3G
          </label>
        </div>
      </div>
      <div
        className="sc-gsTEea ldJggX"
      >
        <div
          style={
            Object {
              "marginBottom": "15px",
            }
          }
        >
          <input
            checked={false}
            className="sc-iBPTik gPelDo"
            id="4g"
            name="internet"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
            value="4g"
          />
          <label
            className="sc-gKseQn gHGxWx"
            htmlFor="4g"
          >
            4G
          </label>
        </div>
      </div>
    </div>
  </div>
  <button
    className="sc-fubCzh kBybUr"
    disabled={false}
    size="default"
    type="submit"
  >
    Submit
  </button>
</form>
`;
